### ISSUE TITLE
Haproxy should reload after configmap-tcp-services was updated

### ISSUE THREAD

**Issue by adamw-linadm** [haproxytech/kubernetes-ingress#421]
Hi I just stared using haproxy as ingress controller. I want to forward multiple tcp services and everything almost behave as expected.
I use --configmap-tcp-services configuration in my controller, but haproxy don't restart after updating config map with tcp services. 
When I update config map with global configuration restart happens automatically.  It only works when I remove config map, wait around 1s and upload it again, then haproxy reloads config with current data from config map.
Is this expected behavior?
Logs:
2022/02/15 15:50:22 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/15 15:56:52 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/15 15:57:53 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/15 15:59:10 INFO    store/events.go:325 configmap 'ingress-controller/haproxy-kubernetes-ingress' updated
2022/02/15 15:59:13 INFO    controller.go:205 HAProxy restarted
[WARNING]  (449) : Exiting Master process...
[NOTICE]   (449) : haproxy version is 2.4.12-4b7772e
[ALERT]    (449) : Current worker #1 (452) exited with code 143 (Terminated)
[WARNING]  (449) : All workers exited. Exiting... (0)
[WARNING]  (490) : config: Can't get version of the global server state file '/var/state/haproxy/global'.
[NOTICE]   (490) : New worker #1 (493) forked
2022/02/15 15:59:48 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/15 16:00:05 INFO    store/events.go:325 configmap 'ingress-controller/haproxy-kubernetes-ingress' updated
2022/02/15 16:00:08 INFO    controller.go:205 HAProxy restarted
[WARNING]  (490) : Exiting Master process...
[NOTICE]   (490) : haproxy version is 2.4.12-4b7772e
[ALERT]    (490) : Current worker #1 (493) exited with code 143 (Terminated)
[WARNING]  (490) : All workers exited. Exiting... (0)
[WARNING]  (511) : config: Can't get version of the global server state file '/var/state/haproxy/global'.
[NOTICE]   (511) : New worker #1 (514) forked

Im using:
NAME   	NAMESPACE         	REVISION	UPDATED                               	STATUS  	CHART                    	APP VERSION
haproxy	ingress-controller	4       	2022-02-15 16:50:39.83672495 +0100 CET	deployed	kubernetes-ingress-1.19.0	1.7.5

**Comment by ivanmatmati** [comment]
Hi @adamw-linadm ,This is what I've done to reproduce your issue. I've started the IC version 1.7.5  with an empty tcp configmap referenced by --configmap-tcp-services command line parameter. Then I insert in data the tcp confguration and I see the correct behavior : 
```
2022/02/17 09:56:28 INFO    store/events.go:325 configmap 'haproxy-controller/haproxy-kubernetes-ingress-tcp' updated
2022/02/17 09:56:29 DEBUG   handler/tcp-services.go:151 TCP frontend 'tcp-32766' created, reload required
2022/02/17 09:56:29 DEBUG   service/service.go:130 Service 'e2e-tests-endpoints/http-echo': new backend 'e2e-tests-endpoints-http-echo-https', reload required
2022/02/17 09:56:29 DEBUG   service/endpoints.go:118 Server slots in backend 'e2e-tests-endpoints-http-echo-https' scaled to match scale-server-slots value: 4, reload required
2022/02/17 09:56:29 DEBUG   maps/main.go:124 Map file 'path-exact' updated, reload required
2022/02/17 09:56:29 DEBUG   maps/main.go:124 Map file 'path-prefix' updated, reload required
2022/02/17 09:56:29 DEBUG   handler/refresh.go:57 Deleting backend 'pprof'
2022/02/17 09:56:29 DEBUG   handler/pprof.go:48 pprof backend created
2022/02/17 09:56:29 DEBUG   maps/main.go:124 Map file 'path-exact' updated, reload required
2022/02/17 09:56:29 DEBUG   maps/main.go:124 Map file 'path-prefix' updated, reload required
2022/02/17 09:56:29 ERROR   process/direct-control.go:65 /tmp/haproxy-ingress/run/haproxy-runtime-api.sock dial unix /tmp/haproxy-ingress/run/haproxy-runtime-api.sock: connect: connection refused
2022/02/17 09:56:29 INFO    controller.go:211 HAProxy reloaded
```
Are you doing differently in your case ?

**Comment by adamw-linadm** [comment]
@ivanmatmati only difference I had one service defined in configmap-tcp-services, and I try to add new one or had 2 services and try to delete one.
```
  kubernetes-ingress-controller:
    Container ID:  containerd://3807065d030adf29784e0629fcdd7e91b87ab2b71a80ccfd00492d044051d9a6
    Image:         haproxytech/kubernetes-ingress:1.7.5
    Image ID:      docker.io/haproxytech/kubernetes-ingress@sha256:47012a34b4faafa8616a746d3abaaee0e81c959bbc28bc8e9b69a44798eeac51
    Ports:         80/TCP, 443/TCP, 1024/TCP
    Host Ports:    80/TCP, 443/TCP, 1024/TCP
    Args:
      --default-ssl-certificate=ingress-controller/haproxy-kubernetes-ingress-default-cert
      --configmap=ingress-controller/haproxy-kubernetes-ingress
      --default-backend-service=ingress-controller/haproxy-kubernetes-ingress-default-backend
      --ingress.class=haproxy
      --publish-service=ingress-controller/haproxy-kubernetes-ingress
      --log=info
      --configmap-tcp-services=ingress-controller/tcp
    State:          Running
      Started:      Thu, 17 Feb 2022 10:25:23 +0100
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     2
      memory:  256Mi
    Requests:
      cpu:      500m
      memory:   128Mi
```

My tcp-services config map looks like:

```apiVersion: v1
kind: ConfigMap
metadata:
  name: tcp
  namespace: ingress-controller
data:
  12000:  # Port where the frontend is going to listen to.
    kube-system/hubble-ui:80 # namespace/srv|pod:port
  11211:
    test/memcached-test:11211
```
So when I remove one entry and upload configmap:

```adam@box ingress]$ git diff tcp-configmaps.yaml
diff --git a/ingress/tcp-configmaps.yaml b/ingress/tcp-configmaps.yaml
index 05dd8c0..112032a 100644
--- a/ingress/tcp-configmaps.yaml
+++ b/ingress/tcp-configmaps.yaml
@@ -4,7 +4,5 @@ metadata:
   name: tcp
   namespace: ingress-controller
 data:
-  12000:  # Port where the frontend is going to listen to.
-    kube-system/hubble-ui:80 # namespace/srv|pod:port
     test/memcached-test:11211
[adam@box ingress]$ kubectl apply -f tcp-configmaps.yaml 
configmap/tcp configured
```
I only have this in haproxy-kubernetes-ingress pods:
```
2022/02/17 09:32:25 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
```
Traffic still can reach port 12000 with defined backend and I still can see entry about this service in haproxy stats UI.
I can remove/add new servicess and nothing changed, until I upload new version of main configmap defined as:
```
--configmap=ingress-controller/haproxy-kubernetes-ingress
```
or restart/kill all haproxy pods.

**Comment by adamw-linadm** [comment]
I think I found problem.. Old version of kubectl on client. I updated version to match server version and reload happend when I added new service.

```
2022/02/17 09:45:31 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/17 09:45:36 INFO    controller.go:211 HAProxy reloaded
[WARNING]  (284) : Reexecuting Master process
[WARNING]  (284) : config: Can't get version of the global server state file '/var/state/haproxy/global'.
[WARNING]  (287) : Proxy healthz stopped (cumulated conns: FE: 1108, BE: 0).
[NOTICE]   (284) : New worker #1 (307) forked
[WARNING]  (287) : Proxy http stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (287) : Proxy https stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (287) : Proxy stats stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (287) : Proxy tcp-11211 stopped (cumulated conns: FE: 86, BE: 0).
[WARNING]  (287) : Proxy tcp-12000 stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (287) : Stopping frontend GLOBAL in 0 ms.
[WARNING]  (287) : Stopping backend ingress-controller-haproxy-kubernetes-ingress-default-backend-http in 0 ms.
[WARNING]  (287) : Stopping backend kube-system-hubble-ui-80 in 0 ms.
[WARNING]  (287) : Stopping backend test-memcached-test-11211 in 0 ms.
```
But not when I remove one.

**Comment by ivanmatmati** [comment]
Still working for me.
```
2022/02/17 11:07:05 INFO    store/events.go:325 configmap 'haproxy-controller/haproxy-kubernetes-ingress-tcp' updated
2022/02/17 11:07:08 DEBUG   handler/tcp-services.go:114 TCP frontend 'tcp-32766' deleted, reload required
```
Will check further if I can reproduce your case but you should consider it works and try to figure out what's going on in your configuration.

**Comment by adamw-linadm** [comment]
@ivanmatmati 
In debug mode:
Adding new serwice:
```
2022/02/18 10:23:04 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/18 10:23:09 DEBUG   handler/tcp-services.go:151 TCP frontend 'tcp-12000' created, reload required
2022/02/18 10:23:09 DEBUG   service/service.go:130 Service 'kube-system/hubble-ui': new backend 'kube-system-hubble-ui-80', reload required
2022/02/18 10:23:09 INFO    controller.go:211 HAProxy reloaded
[WARNING]  (283) : Reexecuting Master process
[WARNING]  (283) : config: Can't get version of the global server state file '/var/state/haproxy/global'.
[WARNING]  (286) : Proxy healthz stopped (cumulated conns: FE: 207, BE: 0).
[NOTICE]   (283) : New worker #1 (308) forked
[WARNING]  (286) : Proxy http stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (286) : Proxy https stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (286) : Proxy stats stopped (cumulated conns: FE: 1, BE: 0).
[WARNING]  (286) : Proxy tcp-11211 stopped (cumulated conns: FE: 94, BE: 0).
[WARNING]  (286) : Proxy tcp-12000 stopped (cumulated conns: FE: 0, BE: 0).
[WARNING]  (286) : Stopping frontend GLOBAL in 0 ms.
[WARNING]  (286) : Stopping backend ingress-controller-haproxy-kubernetes-ingress-default-backend-http in 0 ms.
[WARNING]  (286) : Stopping backend kube-system-hubble-ui-80 in 0 ms.
[WARNING]  (286) : Stopping backend test-memcached-test-11211 in 0 ms.
```

Removing service:
```
2022/02/18 10:23:45 INFO    store/events.go:325 configmap 'ingress-controller/tcp' updated
2022/02/18 10:23:49 DEBUG   handler/tcp-services.go:114 TCP frontend 'tcp-12000' deleted, reload required
2022/02/18 10:23:49 DEBUG   handler/refresh.go:57 Deleting backend 'kube-system-hubble-ui-80'
```
So reload required but not performed. I still can see in stats UI frontend and backend entry for removed service: https://imgur.com/iozrV1U.png

**Comment by Mo3m3n** [comment]
Thanks for your patience @adamw-linadm
I have reproduced this issue and then fixed it in https://github.com/haproxytech/kubernetes-ingress/commit/60f02184968f55340b10c0f52485e00b0c4991ce
It will be backported and we will publish a fixed 1.7 soon, I will let you know.

**Comment by adamw-linadm** [comment]
@Mo3m3n adding/removing tcp-service  now works as it should, thank you very much :)
